### PR TITLE
Count valid emoji as one character when calculating nick length

### PR DESCRIPTION
**Pull request by cpalone**
This loads a list of valid emoji output by the client build process and checks for valid emoji in nicks, counting them as 1 character for the purposes of nick validation.

### CODE REVIEW

File: proto/identity.go [hunk 1]
```diff
@@ -1,10 +1,17 @@
...
 
 import (
+	"encoding/json"
+	"io/ioutil"
+	"regexp"
 	"strings"
 	"unicode/utf8"
 )
 
+var validEmoji = make(map[string]string)
```
Comment (logan) [inline proto/identity.go]: Generally, only use make() if there is no alternative literal constructor (e.g., making a slice of non-zero size or capacity).

```
var validEmoji = map[string]string{}
```

File: proto/identity.go [hunk 2]
```diff
@@ -48,6 +55,35 @@ type IdentityView struct {
...
+}
+
+func nickLen(nick string) int {
+	indices := possibleEmoji.FindAllStringIndex(nick, 36)
+	for _, item := range indices {
+		s := nick[item[0]+1 : item[1]-1]
+		_, ok := validEmoji[s]
+		if ok {
+			ret := 1 + nickLen(nick[0:item[0]])
+			if item[1] < len(nick)-1 {
```
Comment (logan) [inline proto/identity.go]: This undercounts by one if `item[1] == len(nick)-1`, because `item[1]` is the index of the byte following the emoji. Just compare against `len(nick)`. Add a test case for this boundary condition.

File: proto/identity_test.go [hunk 3]
```diff
@@ -81,4 +82,43 @@ func TestNormalizeNick(t *testing.T) {
...
+		name[len(name)-len(":+1:")+2] = '1'
+		name[len(name)-len(":+1:")+3] = ':'
+		expected := "a aa " + string(name[6:])
+		So(pass(string(name)), ShouldEqual, expected)
+	})
+}
+
+func TestNickLen(t *testing.T) {
+	Convey("Length of nicks without emoji are correct", t, func() {
+		validEmoji["greenduck"] = "greenduck"
```
Comment (logan) [inline proto/identity_test.go]: For clarity and convenience, move this to the top level of TestNickLen.

**Comment by logan** [comment]
Ok, this looks good to me! Please squash and then I'll merge.

**Comment by cpalone** [comment]
Done!
